### PR TITLE
feat(core): migrate to Angular 19, fix audit vulnerabilities, and convert components to standalone, replace input output decorators with signal-based inputs and add trivy workflow

### LINKED ISSUE #97: Upgrade to Angular 19, Remove Vulnerabilities, and Migrate to Signals & Standalone Components
This task involves addressing existing security vulnerabilities by upgrading the application to Angular 19. As part of the upgrade, refactor the codebase to adopt Angular Signals for state management and migrate from NgModules to standalone components where applicable.
The goal is to improve security, performance, and long-term maintainability while aligning the project with modern Angular best practices.

### PR DESCRIPTION
**Pull request by piyushsinghgaur1**
This pull request introduces significant updates to modernize the Angular codebase, improve the build and packaging process, and enhance CI/CD workflows. The most notable changes include upgrading Angular and related dependencies to v19, refactoring the custom element bundling script for compatibility with new Angular output formats, updating build configurations, and improving test and workflow reliability.

**Major dependency and build system upgrades:**

- Upgraded Angular framework and related dependencies (including `@angular/*`, `@ng-select/ng-select`, `lerna`, `rxjs`, `zone.js`, etc.) to version 19 in both `package.json` and `projects/workflows-creator/package.json`. This also includes devDependencies and TypeScript updates for compatibility. [[1]](diffhunk://#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519R24-R98) F9c4fafeL9R10, [[2]](diffhunk://#diff-189eecfd586d1a5f723241fe0ba18653826f1ceb7ffd33fa36b371b219a9bcacL17-R52) [[3]](diffhunk://#diff-189eecfd586d1a5f723241fe0ba18653826f1ceb7ffd33fa36b371b219a9bcacL66-R67)
- Updated `angular.json` to use the new Angular v19 builder (`@angular-devkit/build-angular:application`), revised build options for output compatibility, and updated schematic collections. [[1]](diffhunk://#diff-2d2675bb4687601a5c7ccf707455132f90f3516a33716185687e5c41df59ac7dL65-L89) [[2]](diffhunk://#diff-2d2675bb4687601a5c7ccf707455132f90f3516a33716185687e5c41df59ac7dL5-R7) [[3]](diffhunk://#diff-2d2675bb4687601a5c7ccf707455132f90f3516a33716185687e5c41df59ac7dL123-R127)

**Element bundling and packaging improvements:**

- Refactored `bundle-element.ts` to support the new Angular build output structure (with hashed chunk files), improved robustness by dynamically detecting output files, and ensured correct copying of styles and README. The hash calculation and package.json update logic were also improved.
- Adjusted `elementBuild` and `postbuild` scripts in both root and package-level `package.json` files to use the new bundling approach and ensure correct working directories. [[1]](diffhunk://#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519R24-R98) [[2]](diffhunk://#diff-189eecfd586d1a5f723241fe0ba18653826f1ceb7ffd33fa36b371b219a9bcacL17-R52)

**Continuous Integration and workflow enhancements:**

- Added a new GitHub Actions workflow (`trivy.yml`) to run Trivy security scans on pull requests targeting the `master` branch.
- Updated existing GitHub Actions workflows to use newer action versions (`actions/cache@v4`), and changed `npm ci` steps to use `--ignore-scripts` for improved security and reproducibility. [[1]](diffhunk://#diff-a0fe23534b616d51ce686d2a1bcd1a78bc75074aef1a2f6ee96c9469991e1a4cL12-R12) [[2]](diffhunk://#diff-a0fe23534b616d51ce686d2a1bcd1a78bc75074aef1a2f6ee96c9469991e1a4cL24-R24) [[3]](diffhunk://#diff-87db21a973eed4fef5f32b267aa60fcee5cbdf03c67fafdc2a9b553bb0b15f34L40-R40)

**Code and test modernization:**

- Removed `ngx-popperjs` usage from builder tests and templates, replacing popper-based overlays with Angular CDK overlays for better compatibility and maintainability. Updated test mocks and input handling accordingly. [[1]](diffhunk://#diff-bdd01b693549969bc3ce6611a485d3e67309bd7ec99632f9f281f8326f63f708L2-R58) [[2]](diffhunk://#diff-8f3818a781dc71ebc4fb8884b19a9857ad6c9bb936d3f5496d69fdec8591ba9bL9-R10) [[3]](diffhunk://#diff-8f3818a781dc71ebc4fb8884b19a9857ad6c9bb936d3f5496d69fdec8591ba9bL28-R28) [[4]](diffhunk://#diff-8f3818a781dc71ebc4fb8884b19a9857ad6c9bb936d3f5496d69fdec8591ba9bL56-R55) [[5]](diffhunk://#diff-8f3818a781dc71ebc4fb8884b19a9857ad6c9bb936d3f5496d69fdec8591ba9bL68-R76)
- Updated SCSS imports to use the Angular recommended `@use` syntax instead of `@import`.

**Other improvements:**

- Added `overrides` in `package.json` to pin certain transitive dependencies for security and compatibility.

---

### Dependency and Build System Upgrades
- Upgraded all Angular-related dependencies and devDependencies to v19, including TypeScript and related tooling, in both root and project-level `package.json` files. [[1]](diffhunk://#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519R24-R98) F9c4fafeL9R10, [[2]](diffhunk://#diff-189eecfd586d1a5f723241fe0ba18653826f1ceb7ffd33fa36b371b219a9bcacL17-R52) [[3]](diffhunk://#diff-189eecfd586d1a5f723241fe0ba18653826f1ceb7ffd33fa36b371b219a9bcacL66-R67)
- Updated `angular.json` build and schematic configuration to match Angular v19 requirements and output structure. [[1]](diffhunk://#diff-2d2675bb4687601a5c7ccf707455132f90f3516a33716185687e5c41df59ac7dL65-L89) [[2]](diffhunk://#diff-2d2675bb4687601a5c7ccf707455132f90f3516a33716185687e5c41df59ac7dL5-R7) [[3]](diffhunk://#diff-2d2675bb4687601a5c7ccf707455132f90f3516a33716185687e5c41df59ac7dL123-R127)

### Element Bundling and Packaging
- Refactored `bundle-element.ts` to dynamically detect and concatenate Angular build output files, supporting hashed filenames and copying assets robustly.
- Updated build scripts to use the new bundling approach and ensure correct execution context. [[1]](diffhunk://#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519R24-R98) [[2]](diffhunk://#diff-189eecfd586d1a5f723241fe0ba18653826f1ceb7ffd33fa36b371b219a9bcacL17-R52)

### CI/CD and Workflow Improvements
- Added a Trivy security scan workflow for pull requests to `master`.
- Updated GitHub Actions to use `actions/cache@v4` and run `npm ci --ignore-scripts` for improved reliability. [[1]](diffhunk://#diff-a0fe23534b616d51ce686d2a1bcd1a78bc75074aef1a2f6ee96c9469991e1a4cL12-R12) [[2]](diffhunk://#diff-a0fe23534b616d51ce686d2a1bcd1a78bc75074aef1a2f6ee96c9469991e1a4cL24-R24) [[3]](diffhunk://#diff-87db21a973eed4fef5f32b267aa60fcee5cbdf03c67fafdc2a9b553bb0b15f34L40-R40)

### Code and Test Modernization
- Replaced `ngx-popperjs` overlays with Angular CDK overlays in builder components and tests, and updated test mocks for new Angular versions. [[1]](diffhunk://#diff-bdd01b693549969bc3ce6611a485d3e67309bd7ec99632f9f281f8326f63f708L2-R58) [[2]](diffhunk://#diff-8f3818a781dc71ebc4fb8884b19a9857ad6c9bb936d3f5496d69fdec8591ba9bL9-R10) [[3]](diffhunk://#diff-8f3818a781dc71ebc4fb8884b19a9857ad6c9bb936d3f5496d69fdec8591ba9bL28-R28) [[4]](diffhunk://#diff-8f3818a781dc71ebc4fb8884b19a9857ad6c9bb936d3f5496d69fdec8591ba9bL56-R55) [[5]](diffhunk://#diff-8f3818a781dc71ebc4fb8884b19a9857ad6c9bb936d3f5496d69fdec8591ba9bL68-R76)
- Updated SCSS to use the recommended `@use` syntax.

### Security and Dependency Management
- Added `overrides` in `package.json` to pin specific vulnerable or problematic transitive dependencies.

### CODE REVIEW

**Review by rohit-sourcefuse — CHANGES_REQUESTED** [review]
## 🎯 APPROVE WITH CHANGES - Excellent Angular 19 Migration

This is outstanding migration work that demonstrates deep understanding of Angular 19 patterns and modern best practices. The architectural improvements are substantial and well-executed. However, there are several critical issues that need fixing before merge.

## Critical Issues (Must Fix):

### 1. 🔴 Duplicate ngClass Binding
**File:** `projects/workflows-creator/src/lib/builder/group/group.component.html` **Line:** 47

Remove the duplicate `[ngClass]` binding (Angular will only use the last one):

```html
<workflow-node
  [ngClass]="{'last-node': i === group().children.length - 1}"
  [node]="node"
  [ngClass]="{'last-node': i === group().children.length - 1}"  <!-- REMOVE THIS -->
  [isLast]="i === group().children.length - 1"
```

### 2. 🔴 Memory Leak: Unmanaged Overlay Subscriptions
**Files:** `builder.component.ts:459`, `group.component.ts:385,520`

The backdrop click subscriptions will cause memory leaks:

```typescript
// Current problematic code:
this.overlayRef.backdropClick().subscribe(() => this.closeOverlay());

// Fix Option 1: Proper cleanup
private subscription?: Subscription;

ngOnDestroy() {
  this.subscription?.unsubscribe();
}

// Fix Option 2: Auto-complete
this.overlayRef.backdropClick().pipe(take(1)).subscribe(() => this.closeOverlay());
```

### 3. 🔴 Missing Trivy Configuration File
**File:** `.github/workflows/trivy.yml` **Line:** 29

The workflow references a non-existent config file:
```yaml
trivy-config: "${{ github.workspace }}/trivy.yaml"  # File doesn't exist
```

**Fix:** Either add the `trivy.yaml` file or remove this line to use Trivy defaults.

### 4. 🔴 Unused Import Cleanup
**File:** `group.component.spec.ts` **Line:** 37

```typescript
import {popper} from '@popperjs/core';  // Not used - remove this
```

## Important Issues (Should Fix):

### 5. Missing OnDestroy Implementation
Both `BuilderComponent` and `GroupComponent` should implement `OnDestroy`:

```typescript
export class BuilderComponent<E> implements OnInit, OnDestroy {
  ngOnDestroy(): void {
    this.closeOverlay();
  }
}
```

### 6. Effect Error Handling
**File:** `builder.component.ts` **Lines:** 98-127

Consider adding error handling to the signal effect:

```typescript
effect(() => {
  try {
    // existing logic
  } catch (error) {
    console.error('Effect error in BuilderComponent:', error);
    // Handle gracefully - maybe reset state or show user error
  }
});
```

## ✅ Excellent Migration Highlights:

**🎉 Perfect Signal Implementation:**
- Proper use of `input()`, `input.required()`, and `output()`
- Clean migration from `@Input/@Output` decorators
- Correct signal reactivity patterns

**🎉 Outstanding Standalone Components:**
- Well-implemented with proper `imports` arrays
- Clean removal of NgModule dependencies
- Excellent `provideWorkflowCreator()` provider function

**🎉 Solid CDK Integration:**
- Good migration from ngx-popperjs to Angular CDK overlays
- Proper overlay positioning and management
- Clean template updates

**🎉 Modern Build System:**
- Correct Angular 19 builder configuration (`@angular-devkit/build-angular:application`)
- Smart `bundle-element.ts` updates for hashed output files
- Good CI/CD improvements with `--ignore-scripts`

**🎉 Security & Dependencies:**
- Thoughtful dependency pinning with `overrides`
- Good Trivy security scanning addition (once config is fixed)
- Clean SCSS migration to `@use` syntax

## Test Coverage Notes:
Consider adding tests for:
- Signal effect behavior and error handling
- Overlay subscription cleanup verification  
- Component destruction scenarios
- Signal reactivity edge cases

## Minor Cleanup:
- Delete the empty `workflow-builder.module.ts` file entirely

## Overall Assessment:

This migration showcases excellent Angular expertise. The architectural improvements are significant:
- Clean separation with standalone components
- Modern reactive patterns with signals
- Better maintainability with CDK overlays
- Improved build pipeline

**Once the critical issues are addressed (especially memory leaks and duplicate bindings), this will be a stellar migration that sets a great foundation for Angular 19.**

Great work on tackling such a comprehensive framework upgrade! 🚀

**Comment by rohit-sourcefuse** [comment]
## 📋 Quick Action Checklist

For faster resolution, here's a prioritized checklist of the critical fixes:

### Priority 1 - Breaking Issues:
- [x] Remove duplicate `[ngClass]` in `group.component.html` line 47
- [x] Add `trivy.yaml` config file or remove reference in `.github/workflows/trivy.yml` line 29
- [x] Remove unused `popper` import from `group.component.spec.ts` line 37

### Priority 2 - Memory Management:
- [x] Add `OnDestroy` to `BuilderComponent` and `GroupComponent`
- [x] Fix overlay subscription cleanup in `builder.component.ts` line 459
- [x] Fix overlay subscription cleanup in `group.component.ts` lines 385, 520

### Priority 3 - Error Handling:
- [x] Add try-catch to signal effect in `builder.component.ts` lines 98-127
- [x] Add error handling to overlay disposal methods

### Optional Cleanup:
- [x] Delete empty `workflow-builder.module.ts` file
- [ ] Add tests for signal effects and overlay cleanup

The Priority 1 items will cause immediate build/runtime issues, so tackle those first. The memory management fixes (Priority 2) are critical for production stability.

Excellent work on this migration - these are just polish items on an otherwise outstanding Angular 19 upgrade! 🎯

File: projects/workflows-creator/bundle-element.ts
Comment (Deepika516) [inline projects/workflows-creator/bundle-element.ts:1]: please remove the extra comments

Comment (piyushsinghgaur1) [reply projects/workflows-creator/bundle-element.ts:1]: Done

File: projects/workflows-creator/src/lib/types/event.types.spec.ts [hunk 1]
```diff
@@ -26,6 +26,10 @@ describe('YourFile Tests', () => {
       },
     };
     value = {};
+    // Initialize dummy objects for tests
+    workflowEvent = {id: 'we', type: 'test'} as any;
+    workflowAction = {id: 'wa', type: 'test'} as any;
+    node = {id: 'n', name: 'node'} as any;
```
Comment (Deepika516) [inline projects/workflows-creator/src/lib/types/event.types.spec.ts:32]: remove any

File: projects/workflows-creator/src/main.ts [hunk 2]
```diff
@@ -1,9 +1,56 @@
...
-enableProdMode();
+// Create application context for custom element (no component bootstrapping)
+createApplication({
+  providers: [provideWorkflowCreator((window as any).workflowEnv)],
+})
+  .then(appRef => {
+    // Create and register the custom element
+    const webComponent = createCustomElement(BuilderComponent, {
+      injector: appRef.injector,
+    });
```
Comment (Deepika516) [inline projects/workflows-creator/src/main.ts:33]: Remove unnecessary comments. If required, add a single BL comment at the top of the function.

File: projects/workflows-creator/src/test.ts [hunk 3]
```diff
@@ -26,6 +26,25 @@ getTestBed().initTestEnvironment(
 );
 
 // Then we find all the tests.
-const context = require.context('./', true, /\.spec\.ts$/);
-// And load the modules.
-context.keys().map(context);
+const req: any = require as any;
```
Comment (Deepika516) [inline projects/workflows-creator/src/test.ts]: Remove any ,Consider defining a proper interface for the require context or using a more specific type guard to maintain type safety.

If your type can be random key and some value in string format then use Record<string, string>

File: projects/workflows-creator/src/test.ts [hunk 4]
```diff
@@ -26,6 +26,25 @@ getTestBed().initTestEnvironment(
...
+let context: any;
+if (typeof req.context === 'function') {
+  context = req.context('./', true, /\.spec\.ts$/);
```
Comment (Deepika516) [inline projects/workflows-creator/src/test.ts]: remove any and all the unnecessary comments

File: workflow-sandbox/src/app/app.component.scss [hunk 5]
```diff
@@ -0,0 +1,14 @@
+:host {
+  display: block;
+  height: 100%;
+}
+
+::ng-deep {
+  .tree-group.no-node {
```
Comment (Deepika516) [inline workflow-sandbox/src/app/app.component.scss]: don't use ::ng-deep

`encapsulation: ViewEncapsulation.None`

or some other approch

File: workflow-sandbox/src/app/app.component.ts [hunk 6]
```diff
@@ -0,0 +1,254 @@
...
+    properties: new Map(),
+  };
+  _diagram = BASE_XML_VALUE;
+
+  // get diagram(): string {
+  //   return this._diagram;
+  // }
+
+  // set diagram(value: unknown) {
+  //   if (typeof value === 'string' && value.trim().length > 0) {
```
Comment (Deepika516) [inline workflow-sandbox/src/app/app.component.ts]: remove commented code

File: workflow-sandbox/src/app/app.component.ts [hunk 7]
```diff
@@ -0,0 +1,254 @@
...
+
+  onItemChanged(event: any) {
+    this.handleValueChange(event);
+  }
+
+  handleElementClick(event: any) {
+    const selected = event.event ?? event.action;
+
+    switch (selected.getIdentifier()) {
+      case OnIntervalEvent.identifier:
```
Comment (Deepika516) [inline workflow-sandbox/src/app/app.component.ts:150]: remove any

File: workflow-sandbox/src/app/app.component.ts [hunk 8]
```diff
@@ -0,0 +1,254 @@
...
+        break;
+
+      case ChangeColumnValueAction.identifier:
+        selected.state.change('columns', this.NORMALIZED_COLUMN);
+        break;
+    }
+  }
+  private selectedCol: any;
+
+  handleValueChange(event: any) {
```
Comment (Deepika516) [inline workflow-sandbox/src/app/app.component.ts]: remove any

File: projects/workflows-creator/src/main.ts (hunk 2 above)
Comment (piyushsinghgaur1) [reply projects/workflows-creator/src/main.ts:33]: Done

File: projects/workflows-creator/src/test.ts (hunk 3 above)
Comment (piyushsinghgaur1) [reply projects/workflows-creator/src/test.ts]: Done

File: projects/workflows-creator/src/test.ts (hunk 4 above)
Comment (piyushsinghgaur1) [reply projects/workflows-creator/src/test.ts]: done

File: workflow-sandbox/src/app/app.component.scss (hunk 5 above)
Comment (piyushsinghgaur1) [reply workflow-sandbox/src/app/app.component.scss]: done

File: projects/workflows-creator/src/lib/types/event.types.spec.ts (hunk 1 above)
Comment (Deepika516) [reply projects/workflows-creator/src/lib/types/event.types.spec.ts:32]: it used in workflow test cases before also so for now we are not removing any from this spec

**Review by Deepika516 — APPROVED** [review]
